V2.0.0

cat-haines opened the pull request:
Version 2.0.0 of WS2812 class …
- Renamed writePixel(p, color) to set(p, color)
- Renamed writeStrip() to draw()
- Removed clearFrame() and added fill(color, [start], [end])

Updated NeoWeather and README with new methods

cat-haines committed:
Version 2.0.0 of WS2812 class
 - Renamed writePixel(p, color) to set(p, color)
 - Renamed writeStrip() to draw()
 - Removed clearFrame() and added fill(color, [start], [end])

Updated neoweather and README with new methods

cat-haines committed:
Minor README update

cat-haines committed:
More README updates

cat-haines commented:
@ersatzavian - I tested the updated NeoWeather code (and did my best to test), but you please confirm it still plays nice with the new class (NOTE: You'll need to copy/paste the v2.0.0 class as it's not an actual library with a require yet)

bbharris commented on WS2812.class.nut:
```diff
@@ -2,121 +2,125 @@
+    // Sets a pixel in the buffer
+    //   p - the pixel (0 <= p < _frameSize)
+    //   color - [r,g,b] (0 <= r,g,b <= 255)
+    function set(p, color) {
```
"p" should really be "index". It makes it much more consistent with "start" and "end" and technically a pixel has both a location and a value.

bbharris commented on WS2812.class.nut:
```diff
@@ -2,121 +2,125 @@
+    }
+
+
+    // Clears the frame buffer
```
Comments need updating since this is no longer "clear". Probably want a comment which says fill can be used for a clear.

cat-haines committed:
Renamed p/pixel to index for `set` method
 - Updated README
Added note about `draw` method to `fill` documentation

bbharris commented on WS2812.class.nut:
```diff
@@ -2,121 +2,125 @@
+    // Fill the array of representative 1-wire waveforms,
+    // done by the constructor at instantiation
+
+    function _initialize() {
```
Since this is only called by the constructor and should be done exactly once, shouldn't it be inline inside the constructor?

cat-haines committed:
colour -> color
added comments to indicate what each method is replacing

bbharris commented on WS2812.class.nut:
```diff
@@ -2,121 +2,125 @@
+    //   color - [r,g,b] (0 <= r,g,b <= 255)
+    function set(p, color) {
+        assert(p >= 0);
+        assert(p < _frameSize);
```
If we are checking index, shouldn't we also check the color values are 0-255?

bbharris commented on WS2812.class.nut:
```diff
@@ -2,121 +2,125 @@
+    // but does not write it to the pixel strip
+    function fill(color, start=null, end=null) {
+        // Set default values
+        if (start == null) { start = 0; }
```
Should 0 just be the default value in the function line? The only time it matters is if you want to assign an end value but now a start value, then you can call fill(color, null, 5) and get the default behavior for start. Given that 0 is just as easy to write as null, I don't see any advantage here. Likewise for the "end" parameter.

cat-haines replied:
This is probably my recent javascript programming streak bleeding through ;)

I make it so start's default value is 0, but you can't set a default value to a class variable.. so end will need to continue working like this.

bbharris commented on WS2812.class.nut:
```diff
@@ -2,121 +2,125 @@
+    // represent the waveforms to send the numbers 0 to 255. This allows the blobs to be
+    // copied in directly, instead of being built for each pixel - which makes the class faster.
+
+    _bits            = null;
```
Seems like a good use case for STATIC. As is each instance will build an identical 256 byte lookup table

cat-haines committed:
Added initialize logic to constructor since we only ever call it once
Added asserts for 0 >= color >= 255

cat-haines committed:
Updated how we set start's default value in `fill`
- Added comment about why we set end how we do

cat-haines committed:
Made _bits array static, and ensure we only initialize it once

cat-haines commented:
@bbharris - I think I've fixed all the issues you mentioned. Good call on the static _bits array, that makes a lot of sense~!

cat-haines committed:
Moved curly brace for constructor to same line

bbharris commented on WS2812.class.nut:
```diff
@@ -2,121 +2,126 @@
+    // represent the waveforms to send the numbers 0 to 255. This allows the blobs to be
+    // copied in directly, instead of being built for each pixel - which makes the class faster.
+
+    static _bits     = array(256);
```
Should explicitly set the initial value to null: array(256, null)

bbharris commented on WS2812.class.nut:
```diff
@@ -2,121 +2,126 @@
+                _bits[i] = valblob;
+            }
+        }
+
```
Should we have a constructor parameter which allows you to _not_ draw to the leds? In cases where you are displaying something, it would allow you to reboot without the display turning off.

constructor(spiBus, frameSize, draw=true)

or we could drop the "draw()" from the constructor altogether and make draw also return "this" then you can do:
ws <- WS2812(spi, 5).draw();

bbharris commented on WS2812.class.nut:
```diff
@@ -2,121 +2,126 @@
+        _frame.writeblob(_bits[color[1]]);
+        _frame.writeblob(_bits[color[0]]);
+        _frame.writeblob(_bits[color[2]]);
+    }
```
consider returning "this"

cat-haines committed:
Public methods return `this` for easy chaining
Added constructor argument to allow you to not draw on initialization
Updated readme
